flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java

```diff
@@ -160,7 +169,19 @@ public void initializeState(StateInitializationContext context) throws Exception
                 legacyCommitterState.clear();
             }
         }
-        sinkWriter = writerStateHandler.createWriter(initContext, context);
+
+        // use union state to ensure that rescaling works correctly
+        this.endOfInputState =
+                context.getOperatorStateStore().getUnionListState(END_OF_INPUT_STATE_DESC);
```

Comment:
I need help understanding using the state.

I would expect that after receiving EOI we can also not persist anything anymore.